### ISSUE TITLE
IDE0051 should factor in new dependency attributes in .NET 5

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#48206]
**Version Used**: 
3.9.0-1.20480.1+b3c3f4875957e51526000898673291d11708d35b

**Steps to Reproduce**:
```C#
using System;
using System.Diagnostics.CodeAnalysis;
using System.Reflection;

class Program
{
    static void Main() => new Program().Bar();

    [DynamicDependency("Foo")]
    private void Bar() => typeof(Program).GetMethod("Foo", BindingFlags.Instance | BindingFlags.NonPublic).Invoke(this, null);

    private void Foo() => Console.WriteLine("Hello");
}
```

**Expected Behavior**:
.NET 5 adds new attributes that enable dependencies not obvious in the code, e.g. those via reflection, to be used for dead code elimination tools like https://github.com/mono/linker to better understand such dependencies.  IDE0051 is such a tool and should ideally respect these attributes, such that in this example, IDE0051 doesn't fire.

**Actual Behavior**:
IDE0051 fires and suggests deleting the "unused" method:
![image](https://user-images.githubusercontent.com/2642209/94745185-3bc18900-0348-11eb-940e-ccbb2342a9e9.png)

cc: @eerhardt, @nickcraver

**Comment by Youssef1313** [comment]
@stephentoub Couple of questions to better understand this:

- Can the attribute reference a field/method outside of the current class?
- If there are method overloads, what's the correct way to use the attribute to specify a specific overload?
- For Visual Basic, which is case insensitive, should the attribute be case insensitive too?

**Comment by stephentoub** [comment]
> Can the attribute reference a field/method outside of the current class?

Yes.  Even outside of the current assembly.  So, there are certainly places where there's no way IDE0051 could appropriately respect the attribute, and potentially even times when it could theoretically but doing so could be prohibitively expensive.  It'd be good, though, if the appropriate thought could be put into if/where/when/how it would make sense, so as to have a complete rather than disjoint story around this kind of trimming / dead-code removal.

> If there are method overloads, what's the correct way to use the attribute to specify a specific overload?

@marek-safar, @eerhardt, where is the syntax documented?  @Youssef1313, you can use the simple name, in which case it says that all of the overloads should be considered used, or you can name a specific overload by specifying the parameter types, e.g. https://github.com/dotnet/runtime/blob/c44154b15126786260f8a6397248dd04c0e015e3/src/libraries/System.Private.CoreLib/src/System/String.cs#L48

> For Visual Basic, which is case insensitive, should the attribute be case insensitive too?

@marek-safar, what does the linker do here?

**Comment by Youssef1313** [comment]
> It'd be good, though, if the appropriate thought could be put into if/where/when/how it would make sense, so as to have a complete rather than disjoint story around this kind of trimming / dead-code removal.

Probably the IDE team will want to take this to design meeting to decide what are the essential situations to handle and what are the situations that doesn't need to be handled.

**Comment by marek-safar** [comment]
> where is the syntax documented

ILLinker uses the same syntax as Roslyn for documentation so it's documented at https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#processing-the-documentation-file

> For Visual Basic, which is case insensitive, should the attribute be case insensitive too?

Not sure why it should be. ILLinker is IL-based tool and the attribute specifies metadata value.

**Comment by CyrusNajmabadi** [comment]
>  you can use the simple name, in which case it says that all of the overloads should be considered used,

fwiw, that's challenging as i don't think we have such a concept at all in roslyn.  Would it be more appropriate to remove hte 'overload' concept from these strings and say that all dependencies must be individually specified?

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Would this be very similar to what's done in https://github.com/dotnet/roslyn/blob/5b1f31a1326768e46d3ac021cb5001b33955bc8c/src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/AbstractReferenceFinder_GlobalSuppressions.cs? (#44368)

**Comment by huoyaoyuan** [comment]
https://github.com/dotnet/roslyn-analyzers/issues/4243#issuecomment-703520734

**Comment by mavasani** [comment]
Tagging @sbomer who implemented the parser for the string arguments to DynamicDependencyAttribute in https://github.com/mono/linker/pull/1197.

@sbomer What is the best way for Roslyn based tooling to convert the string argument to Roslyn symbol? Given that the parser you added is adapted from Roslyn's DocumentationCommentId parser, would passing the string to [DocumentationCommentId .GetSymbolsForDeclarationId](http://sourceroslyn.io/#Microsoft.CodeAnalysis/DocumentationCommentId.cs,83) cover all valid allowed values?

**Comment by julealgon** [comment]
I would love to be able to leverage this feature. We've had multiple situations where developers deleted methods thinking they were not being used anywhere only to find that they were being called from a WebForms `ObjectDataSource` control via reflection.

I'd like to be able to go and mark those classes/methods to avoid this from happening again.

But I see that the attributes only exist in .NET5 and up. I would hope they would still work for IDE analyzers if the application provides a polyfill though, which could either be manuall or through a project such as this one:
- https://github.com/Sergio0694/PolySharp/
